javascript/ql/src/semmle/javascript/security/dataflow/TaintedPath.qll

```diff
@@ -210,7 +210,7 @@ module TaintedPath {
       exists(DotDotSlashPrefixRemovingReplace call |
         src = call.getInput() and
         dst = call.getOutput() and
-        dstlabel.isAbsolute() and // result can be absolute
+        (srclabel.isNonNormalized() or dstlabel.isAbsolute()) and // if src is normalized, then dst must be absolute (if dst is relative, then dst is sanitized)
```

Comment:
Ok now it's making my head explode 😕

Let's just go over the whole table:
- Normalized relative -> empty set (completely sanitized)
- Raw relative ->
  - Raw relative. Example: `foo/../bar/../..` becomes `foo/bar/../..`
  - Raw absolute. Example: `..//foo/bar/../..` becomes `/foo/bar/../..`
  - Normalized absolute. Example: `..//foo/bar` becomes `/foo/bar`
  - Normalized relative. Example: `..././foo/bar` becomes `../foo/bar`
- Normalized absolute ->
  - Normalized absolute (can't contain `../` so unaffected)
- Raw absolute ->
  - Raw absolute. Example: `/foo/../bar/../..` becomes `/foo/bar/../..`
  - Normalized absolute. Example: `/foo/../bar/` becomes `/foo/bar`.

The line below here doesn't preserve raw->normalized transitions. Missing the `raw relative/absolute -> normalized absolute` in particular can cause FNs for the `FsPathSinkWithoutUpwardNavigation` sink.

Perhaps I'm being a stickler but I'd like it if we can get this transition table exactly right.

Reply:
> Ok now it's making my head explode confused

I also had some trouble conceptualizing what was supposed to happen in each case, and how to express this as compact QL. (I considered basically writing out the entire truth table in QL). 

> Perhaps I'm being a stickler but I'd like it if we can get this transition table exactly right.

It's fine. Thanks for writing out the table.